### PR TITLE
Use JGit 7.0.0, require Java 17 and Jenkins 2.463 or newer

### PR DESCRIPTION
**Pull request by MarkEWaite**
## Use JGit 7.0.0, require Java 17 and Jenkins 2.463 or newer

JGit 7.0.0 requires Java 17.  Jenkins 2.463 requires Java 17.

Use the plugin bill of materials from 2.462.x because it is the closest we have to 2.463.  The plugin is expected to work with any release 2.463 or later.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-client-plugin/blob/master/CONTRIBUTING.adoc) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes

## Types of changes

What types of changes does your code introduce? _Put an `x` in the boxes that apply_

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by olamy — APPROVED** [review]
LGTM Go for it!
But bear in mind Apache Mina SSHD will be released soon (before end of September)
I have done some testing the last few weeks and it was working fine with SNAPSHOT but just in case we need modification after the final release/

**Comment by MarkEWaite** [comment]
This change should have been preceded by a thorough review of the consumers of the JGit API within the Jenkins plugins.  I've run a preliminary review using usage in plugins with the following signatures:

```
org.eclipse.jgit.api.CheckoutCommand#setForce
org.eclipse.jgit.api.RemoteRemoveCommand#setName
org.eclipse.jgit.api.RemoteSetUrlCommand#setName
org.eclipse.jgit.api.RemoteSetUrlCommand#setPush
org.eclipse.jgit.api.RemoteSetUrlCommand#setUri
org.eclipse.jgit.api.StashApplyCommand#setApplyIndex
org.eclipse.jgit.api.StashApplyCommand#setApplyUntracked
org.eclipse.jgit.dircache.DirCacheCheckout#checkoutEntry
org.eclipse.jgit.dircache.DirCacheEntry#getLastModified
org.eclipse.jgit.dircache.DirCacheEntry#mightBeRacilyClean
org.eclipse.jgit.dircache.DirCacheEntry#setLastModified
org.eclipse.jgit.internal.storage.file.FileSnapshot#lastModified
org.eclipse.jgit.internal.storage.file.FileSnapshot#save
org.eclipse.jgit.internal.storage.file.LockFile#getCommmitLastModified
org.eclipse.jgit.lib.CommitBuilder#setEncoding
org.eclipse.jgit.lib.CoreConfig#isLogAllRefUpdates
org.eclipse.jgit.lib.RefDatabase#getRef
org.eclipse.jgit.lib.Repository#hasObject
org.eclipse.jgit.lib.Repository#peel
org.eclipse.jgit.lib.TagBuilder#toByteArray
org.eclipse.jgit.merge.MergeFormatter#formatMerge
org.eclipse.jgit.pgm.ReceivePack#setEchoCommandFailures
org.eclipse.jgit.revwalk.ReachabilityChecker#areAllReachable
org.eclipse.jgit.revwalk.RevWalk#createReachabilityChecker
org.eclipse.jgit.src.org.eclipse.jgit.merge.MergeMessageFormatterformatWithConflicts
org.eclipse.jgit.src.org.eclipse.jgit.patch.PatchApplierapplyPatch
org.eclipse.jgit.storage.file.WindowCacheStats#getOpenBytes
org.eclipse.jgit.storage.file.WindowCacheStats#getOpenFiles
org.eclipse.jgit.transport.ReceivePack.FirstLine
org.eclipse.jgit.transport.RefAdvertiser#send
org.eclipse.jgit.transport.ssh.jsch.JschSession#getSftpChannel
org.eclipse.jgit.transport.Transport#getFilterBlobLimit
org.eclipse.jgit.transport.UploadPack#getFilterBlobLimit
org.eclipse.jgit.transport.UserAgent#getAgent
org.eclipse.jgit.treewalk.FileTreeIterator.Entry#getEntryLastModified
org.eclipse.jgit.treewalk.WorkingTreeIterator.Entry#getEntryLastModified
org.eclipse.jgit.treewalk.WorkingTreeIterator#getEntryLastModified
org.eclipse.jgit.util.FileUtils#lastModified
org.eclipse.jgit.util.FileUtils#setLastModified
org.eclipse.jgit.util.FS#createNewFile
org.eclipse.jgit.util.FS#getLastModifiedTime
org.eclipse.jgit.util.FS#lastModified
org.eclipse.jgit.util.FS#setAsyncFileStoreAttributes
org.eclipse.jgit.util.FS#setLastModified
org.eclipse.jgit.util.io.AutoLFInputStream
```

That set of removed method signatures leads to the following detected issues:

* [JENKINS-73775](https://issues.jenkins.io/browse/JENKINS-73775) - Blue ocean git commit from Pipeline editor fails with JGit method not found exception
* [JENKINS-73781](https://issues.jenkins.io/browse/JENKINS-73781) - last changes plugin fails with peel() method not found exception
* [JENKINS-73780](https://issues.jenkins.io/browse/JENKINS-73780) - maven release cascade plugin fails SCM checkout with setForce(boolean) method not found exception

**Comment by railsharipov** [comment]
Looks like these changes break `Git Changelog` plugin:
```
java.lang.IllegalAccessError: class se.bjurr.gitchangelog.internal.git.GitRepo tried to access private method 'org.eclipse.jgit.lib.Ref org.eclipse.jgit.lib.Repository.peel(org.eclipse.jgit.lib.Ref)' (se.bjurr.gitchangelog.internal.git.GitRepo is in unnamed module of loader 'PluginClassLoader for git-changelog' @568f4cfa; org.eclipse.jgit.lib.Repository is in unnamed module of loader 'PluginClassLoader for git-client' @479a864c)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.getPeeledObjectId(GitRepo.java:171)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.findRef(GitRepo.java:164)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.getRef(GitRepo.java:149)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.firstCommit(GitRepo.java:242)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.getCommit(GitRepo.java:104)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getFrom(GitChangelogApi.java:776)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:717)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:90)
 	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:84)
 	at PluginClassLoader for git-changelog//org.jenkinsci.plugins.gitchangelog.steps.GitChangelogStep.perform(GitChangelogStep.java:451)
 	at PluginClassLoader for git-changelog//org.jenkinsci.plugins.gitchangelog.steps.GitChangelogStep$1$1.call(GitChangelogStep.java:353)
 	at hudson.remoting.UserRequest.perform(UserRequest.java:225)
 	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
 	at hudson.remoting.Request$2.run(Request.java:391)
 	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:81)
 	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
 	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:140)
 	at java.base/java.lang.Thread.run(Unknown Source)
 Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: dc8858c2-7af4-492c-afea-45b0b9b25d4b
 Caused: java.io.IOException: Remote call on JNLP4-connect connection from ip-10-0-102-77.ec2.internal/10.0.102.77:56808 failed
 	at hudson.remoting.Channel.call(Channel.java:1116)
 	at hudson.FilePath.act(FilePath.java:1319)
 	at PluginClassLoader for git-changelog//org.jenkinsci.plugins.gitchangelog.steps.GitChangelogStep$1.run(GitChangelogStep.java:357)
 	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
 	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
 	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by MarkEWaite** [comment]
> Looks like these changes break `Git Changelog` plugin:
> 
> ```
> java.lang.IllegalAccessError: class se.bjurr.gitchangelog.internal.git.GitRepo tried to access private method 'org.eclipse.jgit.lib.Ref org.eclipse.jgit.lib.Repository.peel(org.eclipse.jgit.lib.Ref)' (se.bjurr.gitchangelog.internal.git.GitRepo is in unnamed module of loader 'PluginClassLoader for git-changelog' @568f4cfa; org.eclipse.jgit.lib.Repository is in unnamed module of loader 'PluginClassLoader for git-client' @479a864c)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.getPeeledObjectId(GitRepo.java:171)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.findRef(GitRepo.java:164)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.getRef(GitRepo.java:149)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.firstCommit(GitRepo.java:242)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.internal.git.GitRepo.getCommit(GitRepo.java:104)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getFrom(GitChangelogApi.java:776)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:717)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:90)
>  	at PluginClassLoader for git-changelog//se.bjurr.gitchangelog.api.GitChangelogApi.getChangelog(GitChangelogApi.java:84)
>  	at PluginClassLoader for git-changelog//org.jenkinsci.plugins.gitchangelog.steps.GitChangelogStep.perform(GitChangelogStep.java:451)
>  	at PluginClassLoader for git-changelog//org.jenkinsci.plugins.gitchangelog.steps.GitChangelogStep$1$1.call(GitChangelogStep.java:353)
>  	at hudson.remoting.UserRequest.perform(UserRequest.java:225)
>  	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
>  	at hudson.remoting.Request$2.run(Request.java:391)
>  	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:81)
>  	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
>  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
>  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
>  	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:140)
>  	at java.base/java.lang.Thread.run(Unknown Source)
>  Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: dc8858c2-7af4-492c-afea-45b0b9b25d4b
>  Caused: java.io.IOException: Remote call on JNLP4-connect connection from ip-10-0-102-77.ec2.internal/10.0.102.77:56808 failed
>  	at hudson.remoting.Channel.call(Channel.java:1116)
>  	at hudson.FilePath.act(FilePath.java:1319)
>  	at PluginClassLoader for git-changelog//org.jenkinsci.plugins.gitchangelog.steps.GitChangelogStep$1.run(GitChangelogStep.java:357)
>  	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
>  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
>  	at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
>  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
>  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
>  	at java.base/java.lang.Thread.run(Unknown Source)
> ```

Unfortunately, there is not much that I can do about that issue.  The class `se.bjurr.gitchangelog.internal.git.GitRepo` in the `se.bjurr.gitchangelog.git-changelog-lib` library has made the same (flawed) decision that the git client plugin made.  It has ignored the deprecation warnings from earlier JGit releases and will now need to be modified to adapt to the APIs that have been removed from JGit 7.0.0.

The changes in the git client plugin to adapt to JGit 7.0.0 were not large, but they did require changes.

**Comment by MarkEWaite** [comment]
> Looks like these changes break `Git Changelog` plugin:

Git changelog plugin has updated the dependent library and released a new version.

**Comment by railsharipov** [comment]
I confirm that issue with `Git Changelog` plugin has been resolved. Thank you, @MarkEWaite !

File: pom.xml [hunk 1]
```diff
@@ -58,15 +58,19 @@
...
     <changelist>-SNAPSHOT</changelist>
     <!-- Character set tests fail unless file.encoding is set -->
     <argLine>-Dfile.encoding=${project.build.sourceEncoding}</argLine>
     <gitHubRepo>jenkinsci/${project.artifactId}-plugin</gitHubRepo>
     <!-- https://www.jenkins.io/doc/developer/plugin-development/choosing-jenkins-baseline/ -->
-    <jenkins.baseline>2.440</jenkins.baseline>
-    <jenkins.version>${jenkins.baseline}.3</jenkins.version>
-    <jgit.version>6.10.0.202406032230-r</jgit.version>
+    <jenkins.baseline>2.462</jenkins.baseline>
+    <!-- TODO Replace with the standard jenkins.baseline references after LTS requires Java 17  -->
```
Comment (basil) [inline pom.xml:68]: This TODO can now be completed.

Comment (MarkEWaite) [reply pom.xml:68]: Thanks.  Completed in:

* #1209

**Pull request merged**